pkg/ingester/wal.go

```diff
@@ -936,3 +1060,109 @@ func SegmentRange(dir string) (int, int, error) {
 	}
 	return first, last, nil
 }
+
+func decodeCheckpointRecord(rec []byte, m proto.Message) (_ proto.Message, err error) {
+	switch RecordType(rec[0]) {
+	case CheckpointRecordType1:
+		if err := proto.Unmarshal(rec[1:], m); err != nil {
+			return m, err
+		}
+	default:
+		// The legacy proto record will have it's first byte >7.
+		// Hence it does not match any of the existing record types.
+		err := proto.Unmarshal(rec, m)
+		if err != nil {
+			return m, err
+		}
+	}
+
+	return m, err
+}
+
+func encodeWithTypeHeader(m proto.Message, typ RecordType, b []byte) ([]byte, error) {
+	buf, err := proto.Marshal(m)
+	if err != nil {
+		return b, err
+	}
+
+	b = append(b[:0], byte(typ))
+	b = append(b, buf...)
+	return b, nil
+}
+
+// WALRecord is a struct combining the series and samples record.
+type WALRecord struct {
+	UserID  string
+	Series  []tsdb_record.RefSeries
+	Samples []tsdb_record.RefSample
+}
+
+func (record *WALRecord) encodeSeries(b []byte) []byte {
+	buf := encoding.Encbuf{B: b}
+	buf.PutByte(byte(WALRecordSeriesType1))
+	buf.PutUvarintStr(record.UserID)
+
+	var enc tsdb_record.Encoder
+	// The 'encoded' already has the type header and userID here, hence re-using
+	// the remaining part of the slice (i.e. encoded[len(encoded):])) to encode the series.
+	encoded := buf.Get()
+	encoded = append(encoded, enc.Series(record.Series, encoded[len(encoded):])...)
+
+	return encoded
+}
+
+func (record *WALRecord) encodeSamples(b []byte) []byte {
+	buf := encoding.Encbuf{B: b}
+	buf.PutByte(byte(WALRecordSamplesType1))
+	buf.PutUvarintStr(record.UserID)
+
+	var enc tsdb_record.Encoder
+	// The 'encoded' already has the type header and userID here, hence re-using
+	// the remaining part of the slice (i.e. encoded[len(encoded):]))to encode the samples.
+	encoded := buf.Get()
+	encoded = append(encoded, enc.Samples(record.Samples, encoded[len(encoded):])...)
+
+	return encoded
+}
+
+func decodeWALRecord(b []byte, rec *Record, walRec *WALRecord) (_ *Record, _ *WALRecord, err error) {
```

Comment:
Why does this function return `*Record` and `*WALRecord` when it simply returns input parameters for these return values?

Reply:
You are right, this is not needed